core/src/main/java/io/grpc/internal/ManagedChannelImpl.java

```diff
@@ -145,15 +147,36 @@
       this.executor = executor;
     }
     this.backoffPolicyProvider = backoffPolicyProvider;
+    this.nameResolver = getNameResolver(target, nameResolverFactory, nameResolverParams);
+    this.loadBalancer = loadBalancerFactory.newLoadBalancer(nameResolver.getServiceAuthority(), tm);
+    this.transportFactory = transportFactory;
+    this.userAgent = userAgent;
+    this.interceptorChannel = ClientInterceptors.intercept(new RealChannel(), interceptors);
+    scheduledExecutor = SharedResourceHolder.get(TIMER_SERVICE);
+
+    this.nameResolver.start(new NameResolver.Listener() {
+      @Override
+      public void onUpdate(List<ResolvedServerInfo> servers, Attributes config) {
+        loadBalancer.handleResolvedAddresses(servers, config);
+      }
+
+      @Override
+      public void onError(Status error) {
+        Preconditions.checkArgument(!error.isOk(), "the error status must not be OK");
+        loadBalancer.handleNameResolutionError(error);
+      }
+    });
+  }
 
+  @VisibleForTesting
+  static NameResolver getNameResolver(String target, NameResolver.Factory nameResolverFactory,
+      Attributes nameResolverParams) {
     // Finding a NameResolver. Try using the target string as the URI. If that fails, try prepending
     // "dns:///".
-    NameResolver nameResolver = null;
-    URI targetUri;
+    URI targetUri = null;
     StringBuilder uriSyntaxErrors = new StringBuilder();
     try {
       targetUri = new URI(target);
-      nameResolver = nameResolverFactory.newNameResolver(targetUri, nameResolverParams);
       // For "localhost:8080" this would likely return null, because "localhost" is parsed as the
```

Comment:
Please make the comment more accurate. "this would likely return null" -> "this method would likely return null"